packages/core/src/sync-realtime/service.ts

```diff
@@ -202,444 +219,506 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
+  private syncBatch = async (newBlock: RealtimeBlock) => {
+    const newBlockNumber = hexToNumber(newBlock.number);
+    const newBlockTimestamp = hexToNumber(newBlock.timestamp);
 
-        this.finalizedBlockNumber = newFinalizedBlock.number;
+    // Get logs
+    const logs = await this._eth_getLogs({
+      fromBlock: numberToHex(this.mostRecentBlock.number + 1),
+      toBlock: numberToHex(newBlockNumber),
+    });
 
-        this.emit("finalityCheckpoint", {
-          blockTimestamp: newFinalizedBlock.timestamp,
-          chainId: this.network.chainId,
-          blockNumber: newFinalizedBlock.number,
-        });
+    const matchedLogs = await this.getMatchedLogs(
+      logs,
+      BigInt(newBlockNumber),
+      true,
+    );
 
-        this.common.logger.debug({
-          service: "realtime",
-          msg: `Updated finality checkpoint to ${newFinalizedBlock.number} (network=${this.network.name})`,
-        });
-      }
+    // Get blocks
+    const missingBlockNumbers = dedupe(
+      matchedLogs.map((log) => log.blockNumber!),
+    ).filter((b) => b !== newBlock.number);
 
-      this.common.logger.debug({
-        service: "realtime",
-        msg: `Finished syncing new head block ${newBlock.number} (network=${this.network.name})`,
-      });
+    const blocks = await Promise.all(
+      missingBlockNumbers.map(this._eth_getBlockByNumber),
+    );
+    blocks.push(newBlock);
 
-      return;
-    }
+    // Insert logs + blocks + transactions
+    await this.insertRealtimeBlocks(matchedLogs, blocks);
 
-    // 3) At least one block is missing.
-    // Note that this is the happy path for the first task after setup, because
-    // the unfinalized block range must be fetched (eg 32 blocks on mainnet).
-    if (newBlock.number > previousHeadBlock.number + 1) {
-      const missingBlockNumbers = range(
-        previousHeadBlock.number + 1,
-        newBlock.number,
-      );
+    this.emit("realtimeCheckpoint", {
+      blockTimestamp: newBlockTimestamp,
+      chainId: this.network.chainId,
+      blockNumber: newBlockNumber,
+    });
 
-      // Fetch all missing blocks using a request concurrency limit of 10.
-      const limit = pLimit(10);
+    this.common.metrics.ponder_realtime_latest_block_number.set(
+      { network: this.network.name },
+      newBlockNumber,
+    );
+    this.common.metrics.ponder_realtime_latest_block_timestamp.set(
+      { network: this.network.name },
+      newBlockTimestamp,
+    );
 
-      const missingBlockRequests = missingBlockNumbers.map((number) => {
-        return limit(async () => {
-          const block = await this.requestQueue.request({
-            method: "eth_getBlockByNumber",
-            params: [numberToHex(number), true],
-          });
-          if (!block) {
-            throw new Error(`Failed to fetch block number: ${number}`);
-          }
-          return block as BlockWithTransactions;
-        });
+    this.mostRecentBlock = realtimeBlockToLightBlock(newBlock);
```

Comment:
Any reason not to call this `latestBlock`?

Reply:
At first it was confusing because there was also a function called "getLatestBlock". But now I think it makes sense.